chain/car_test.go

```diff
@@ -0,0 +1,190 @@
+			msgs,
+			rcts,
+		)
+	})
+
+	var buf bytes.Buffer
+	carW := bufio.NewWriter(&buf)
+
+	// export the car file to a buffer
+	exportedKey, err := chain.Export(ctx, ts11, cb, cb, carW)
+	assert.NoError(t, err)
+	assert.Equal(t, ts11.Key(), exportedKey)
+	require.NoError(t, carW.Flush())
+
+	mds := ds.NewMapDatastore()
+	bstore := blockstore.NewBlockstore(mds)
+
+	// import the car file from the buffer
+	carR := bufio.NewReader(&buf)
+	importedKey, err := chain.Import(ctx, bstore, carR)
+	assert.NoError(t, err)
+	assert.Equal(t, ts11.Key(), importedKey)
+
+	// walk the blockstore and assert it had all blocks imported
+	validateBlockstoreImport(t, ts11.Key(), gene.Key(), bstore)
+}
+
+func TestChainImportExportMultiTipSetWithMessages(t *testing.T) {
+	tf.UnitTest(t)
+	ctx := context.Background()
+
+	keys := types.MustGenerateKeyInfo(1, 42)
+	mm := types.NewMessageMaker(t, keys)
+	alice := mm.Addresses()[0]
+	cb := chain.NewBuilder(t, address.Undef)
+
+	gene := cb.NewGenesis()
+	ts10 := cb.AppendManyOn(9, gene)
+	msgs := []*types.SignedMessage{
+		mm.NewSignedMessage(alice, 1),
+		mm.NewSignedMessage(alice, 2),
+		mm.NewSignedMessage(alice, 3),
+		mm.NewSignedMessage(alice, 4),
+		mm.NewSignedMessage(alice, 5),
+	}
+	rcts := types.EmptyReceipts(5)
+	ts11 := cb.BuildOneOn(ts10, func(b *chain.BlockBuilder) {
+		b.AddMessages(
+			msgs,
+			rcts,
+		)
+	})
+
+	ts12 := cb.AppendOn(ts11, 3)
+
+	var buf bytes.Buffer
+	carW := bufio.NewWriter(&buf)
+
+	// export the car file to a buffer
+	exportedKey, err := chain.Export(ctx, ts12, cb, cb, carW)
+	assert.NoError(t, err)
+	assert.Equal(t, ts12.Key(), exportedKey)
+	require.NoError(t, carW.Flush())
+
+	mds := ds.NewMapDatastore()
+	bstore := blockstore.NewBlockstore(mds)
+
+	// import the car file from the buffer
+	carR := bufio.NewReader(&buf)
+	importedKey, err := chain.Import(ctx, bstore, carR)
+	assert.NoError(t, err)
+	assert.Equal(t, ts12.Key(), importedKey)
+
+	// walk the blockstore and assert it had all blocks imported
+	validateBlockstoreImport(t, ts12.Key(), gene.Key(), bstore)
+}
+
+func validateBlockstoreImport(t *testing.T, start, stop types.TipSetKey, bstore blockstore.Blockstore) {
+	// walk the blockstore and assert it had all blocks imported
+	cur := start
+	for !cur.Equals(stop) {
+		parents := cid.NewSet()
+		for _, c := range cur.ToSlice() {
+			bsBlk, err := bstore.Get(c)
+			assert.NoError(t, err)
+			blk, err := types.DecodeBlock(bsBlk.RawData())
+			assert.NoError(t, err)
+
+			if !blk.Messages.Equals(types.EmptyMessagesCID) {
+				bsMsgs, err := bstore.Get(blk.Messages)
+				assert.NoError(t, err)
+				_, err = types.DecodeMessages(bsMsgs.RawData())
+				assert.NoError(t, err)
+			}
+
+			if !blk.MessageReceipts.Equals(types.EmptyReceiptsCID) {
+				bsRcts, err := bstore.Get(blk.MessageReceipts)
+				assert.NoError(t, err)
+				_, err = types.DecodeReceipts(bsRcts.RawData())
+				assert.NoError(t, err)
+			}
+
+			for _, p := range blk.Parents.ToSlice() {
+				parents.Add(p)
+			}
```

Comment:
The CidSet is unnecessary, just use a slice. All blocks in a tipset should have the same parent tipset key (you can require this).

Reply:
addressed.